docs: add a design doc for TokenAirdropTransaction

IvanKavaldzhiev opened the pull request:
**Description**:
<!--
One or two line summary of what this PR does and why it is needed, followed by a list
of changes in imperative, present tense for use in the commit message or changelog. Example:

This PR modifies ... in order to support ...
* Add config property
* Change column name
* Remove ...
-->

**Related issue(s)**:

Fixes #

**Notes for reviewer**:
<!-- Provide logs, performance numbers or screenshots of the new functionality -->

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+The airdrop transaction would include several types of fees, a different set of them would be applicable, based on the conditions the airdrop is performed to. The list of fees is:
+
+- Airdrop fee covering the operation itself
+- Association fee covering a pre-paid fee for associating the token being airdropped to the receiver, in case there is a missing relationship between the two
```
This is only if the `maxAutoAssociations` is not greater than `usedAutoAssociations` right?

IvanKavaldzhiev replied:
This association fee is not related to auto associations, but for explicit association which would happen in case the airdropped is claimed by the receiver. Since we don't want to add this expense to the receiving side, the airdrop sender should pay in advance for this potential future association.

If there are free auto association slots for the recipient, the airdrop would directly be transferred to the recipient and I think no association fee should be paid in this case.

jsync-swirlds replied:
Based on the HIP the sender will pay for the automatic association for a recipient who has an automatic association slot available.
>As an hts-token-airdrop-sender I want to cover the fees (custom, association & rent until the end of the first full auto renewal period) of an airdrop that consumes an automatic token slot to reduce the friction for a receiving account

IvanKavaldzhiev replied:
Oh, I thought automatic associations are not covered by fees and only a transfer fee is paid in those cases. If they are, this association fee would also cover this auto association step, only if `maxAutoAssociations` >= `usedAutoAssociations`, so there are free auto association slots available.

jsync-swirlds replied:
Interesting, I had understood that automatic association slots had a cost (both a fee when consumed, and a "rent" when the account is renewed).

Thanks for clarifying.

IvanKavaldzhiev replied:
Actually in the diagram in the HIP we can see that even if the receiver has free auto associate slots, the airdrop sender should still cover an association fee, so I think @jsync-swirlds is right and the association fee charged during an airdrop covers both a direct auto association or a potential manually triggered association, during a claim by the recipient. In any case, we should charge this association fee from the airdrop sender. That's why we also change the existing `CryptoTransfer` fee logic by making the sender always paying for the token association.

@Nana-EC , @Neeharika-Sompalli , what do you think?

<img width="1297" alt="image" src="https://github.com/hashgraph/hedera-services/assets/5901140/0a07bdb8-fa2c-42dd-91a6-3863cfe1c3ef">

Neeharika-Sompalli replied:
Ah I see .. so, we don't charge `maxAutoAssociations` during creation anymore. Then this makes sense.

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+        - Check that the airdrop sender account is a valid account. That is:
+          - an existing account, which was not deleted or expired
+          - a non-existing account, pointed by an alias of `evm_address` or public `ECDSA key`
+        - Check that the token being airdropped, does not contain a custom fee, which needs to be covered by the recipient. Such transactions should be reverted and rejected.
```
You can mention `fallbackRoyaltyFee` here also

IvanKavaldzhiev replied:
Noted

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a `TokenAssociation`
+
+              - It is an existing one and has `max_auto_associations`, which is a positive number and there are free auto association slots and the recipient is not associated to the token →
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a `TokenAssociation`
```
Need to be clear here that it is added to `auto_associations` section of `TransactionRecord`

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+              - It is an existing one and has `max_auto_associations`, which is a positive number and there are free auto association slots and the recipient is not associated to the token →
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a `TokenAssociation`
+
+              - It is an existing one and has `max_auto_associations`, which is a positive number but there are no free auto association slots and the recipient is not associated to the token →
+                the token transfer is interpreted as a pending and a `TokenPendingAirdrop` entry is added to the `TransactionRecord`
```
Here it will be added only to `pendingAirdrops` section in `TransactionRecord` right ?

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+        - The business logic for sending pending airdrops
+            - Check the association status and the existence of the recipient:
+
+              - It is an existing one and has `max_auto_associations` set to -1 and is not associated to the token →
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a `TokenAssociation`
```
Need to be clear here that it is added to auto_associations section of TransactionRecord

IvanKavaldzhiev replied:
Noted

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+        - assess all required fees for the airdrop from the payer of the transaction. This includes:
+            - airdrop fee (assessed always)
+            - first auto-renew fee for the token (assessed always)
+            - custom token fees covered by the sender (assessed always if present)
```
Do we store these custom fees somewhere? When are the custom fees charged ?

IvanKavaldzhiev replied:
I propose not to store them. They would be charged in the `TokenAirdrop.calculateFees` during the sending of airdrop

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+## Purpose
+
+The essence of HIP-904 is introducing a missing functionality in the Hedera ecosystem that enables airdrop transactions. 
+Airdrops are frequently used operation in the web3 ecosystem and a lot of blockchain, dapp or DAO developers incentivise users to explore and use their products through it. Early adopters are ofter awarded with the so called airdrops which are native tokens for the system and a dedicated amount of them are distributed in airdrops. Hedera could benefit from introducing this functionality and this  ...
```
This reads better with some line wrapping (which will be removed in the rendered view), and there are a couple misspelled words and awkward grammar.
```suggestion
Airdrops are frequently used operation in the web3 ecosystem and a lot of blockchains,
 dapps, or DAO developers incentivize users to explore and use their products with
 this process. Early adopters are often awarded with so-called airdrops which are
 tokens for the system and a dedicated amount of them are distributed in such airdrops.
 Hedera will benefit from introducing this functionality and this is something long
 requested by the community.
```

IvanKavaldzhiev replied:
Noted

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+
+### Fee types
+
+The airdrop transaction would include several types of fees, a different set of them would be applicable, based on the conditions the airdrop is performed to. The list of fees is:
```
Misuse of `would` here, and a slight improvement in wording.
```suggestion
The airdrop transaction will include several types of fees, a different set of them will be applicable, based on the characteristics of the airdrop. The list of fees is:
```

IvanKavaldzhiev replied:
Noted

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+- Auto-renewal fees covering a pre-paid period for a full auto-renewal cycle
+- Fee covering auto-creation/lazy-creation of accounts, if we airdrop to an alias address of non-existing account
+
+Note that for all these fees, a given sender, which is the actual token owner, could be covered by a different payer, which is willing to pay. However, in case the receiver has not claimed an airdrop for a more than a full auto-renew cycle, the sender (token owner) would be required to continue paying future rents and they cannot be covered by a separate payer.
```
I would recommend leaving off the last clause ` and they cannot be covered by a separate payer`.  Nothing in the HIP, that I can see, requires that nobody can pay those fees on the sender's behalf; in fact it is not unreasonable, in my opinion, to allow an account identified in the `auto_renew_account_id` for the sender to cover those rent fees (although that is not required by this HIP either).

IvanKavaldzhiev replied:
I agree with you. This statements makes the doc a bit restrictive and this might extend in the future. I removed the last clause.

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+
+If the sender of the airdrop is also the payer of this transaction, we would need only their key to be present in the transaction context.
+
+If the airdrop transaction fees from the list above are covered by a separate payer, then both the keys of the airdrop sender and the payer should be present in the context. That is needed, so that the sender authorizes giving an airdrop from their own balance.
```
```suggestion
If the airdrop transaction fees from the list above are covered by a separate payer, then the signatures of both the airdrop sender and the payer must be present in the context. This is so because the sender is required to authorize sending the airdrop tokens from their balance.
```

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+- Update `TokenServiceDefinition` class to include the new RPC method definition for sending airdrops
+- Implement new `TokenAirdropHandler` class which should be invoked when the gRPC server handles `TokenAirdropTransaction` transactions. The class should be responsible for:
+    - Verify that the pending airdrops list contains between 1 and 10 entries, inclusive
+    - Verify that the pending airdrops list does not have any duplicate entries
```
Do we disbar duplicates?  What if an airdrop smart contract were to send 3 tokens from Bob to Alice for using a new dApp, and send a bonus 3 tokens from Bob to Alice for "liking" the dApp account on `X`?

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+        - Any additional validation depending on config or state i.e. semantics checks
+        - Check that the airdrop sender account is a valid account. That is:
+          - an existing account, which was not deleted or expired
+          - a non-existing account, pointed by an alias of `evm_address` or public `ECDSA key`
```
I do not think this specific line would be valid, because the non-existent account would not have any balance to send...

IvanKavaldzhiev replied:
You're correct! My bad, I was probably referring to the recipient accounts, but this specific section is for the sender accounts only. I'll delete the second case.

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+              - It is an existing one and has `max_auto_associations`, which is a positive number and there are free auto association slots and the recipient is not associated to the token →
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a new `TokenAssociation` entry in the `automatic_token_associations`
+
+              - It is an existing one and has `max_auto_associations`, which is a positive number but there are no free auto association slots and the recipient is not associated to the token →
+                the token transfer is interpreted as a pending and a `TokenPendingAirdrop` entry is added to the `TransactionRecord`
```
Do we need to specify also that the PendingAirdropId/PendingAirdropValue entry is added to the PendingAirdrop virtual map in state?

IvanKavaldzhiev replied:
I added this clarification.

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+        - Such transactions should be reverted and rejected.
+        - The business logic for sending pending airdrops
+            - Check the association status and the existence of the recipient:
+
```
Probably shouldn't have blank lines between bullet points.  Some markdown processing will create independent lists in that case.

IvanKavaldzhiev replied:
Noted

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+        - The business logic for sending pending airdrops
+            - Check the association status and the existence of the recipient:
+
+              - It is an existing one and has `max_auto_associations` set to -1 and is not associated to the token →
+                the token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a new `TokenAssociation` entry in the `automatic_token_associations`
```
It looks like you're trying to create a case/answer structure with the arrow at end of line, if that's the case, you probably need to add an actual line break ("view" this file to see how github removes the line breaks otherwise).
```suggestion
              - It is an existing one and has `max_auto_associations` set to -1 and is not associated to the token →<br/>
                The token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a new `TokenAssociation` entry in the `automatic_token_associations`
```

IvanKavaldzhiev replied:
Noted

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+* Verify that an airdrop with a token and a missing recipient pointed by its public `ECDSA key alias`, works correctly by auto-creating the recipient with `maxAutoAssociations=-1` and the token is directly transferred to it
+* Verify that an airdrop with a token and a missing recipient pointed by its `evm_address` alias, works correctly by lazy-creating the recipient with `maxAutoAssociations=-1` and the token is directly transferred to it
+* Verify that an airdrop with a missing TokenID fails
+* Verify that an airdrop with a missing recipient pointed by an `ED25519 key alias` fails
```
Are we certain this is not supposed to work?  As far as I know it is entirely valid to send a `CryptoTransfer` to an ED25519 public key and it will auto-create the account, setting the `alias` to the public key bytes.

IvanKavaldzhiev replied:
You're right. That's still a valid case. I removed this scenario.

IvanKavaldzhiev commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,208 @@
+ * pending airdrop.
+ *
+ * Any airdrop that completes immediately SHALL be irreversible. Any airdrop that results in a
+ * "Pending Airdrop" MAY be canceled via a `cancelAirdrop` transaction.<
```
Noted

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,205 @@
+     * a credit, so each _fungible_ token transfer MUST have _balanced_ entries in the
+     * TokenTransferList for that transfer.</blockquote>
+     */
+     repeated TokenTransferList token_transfers= 1;
```
Using a `TokenTransferList` here doesn't seem right. How are we doing to enforce there is one credit and one debit in each transfer ? 

Seems like we need a different protobuf definition that has one credit and one debit.
Allowing user to set multiple credits and debits will make it confusing to create a `PendingAirdropId` that needs one sender and one receiver

Neeharika-Sompalli replied:
@tinker-michaelj @netopyr @Nana-EC   What do you think ?

jsync-swirlds replied:
Each Pending Airdrop is created individually from a _single_ transfer in the list, and only if _necessary_. The majority of transfers are expected to complete without creating a pending airdrop.
I would expect to see one PendingAirdropId for _each_ combination of sender and recipient in the transfer list(s), and _only_ if the recipient cannot immediately receive the airdrop.

IvanKavaldzhiev replied:
In the case of fungible airdrops having the same sender and receiver - we would result in a single transfer with the fungible value aggregated.

In the case of NFT airdrops with different serial numbers but the same sender and receiver - we would have a single transfer in the list for each separate serial number with the same sender and receiver.

This is my understanding, having the current HIP definition in mind.

Neeharika-Sompalli replied:
I still don't understand, I think we will need to add validations in code. But probably don't want to block this PR. 🙂

Neeharika-Sompalli reviewed (dismissed):
LGTM ! But I am not sure adding `TokenTransferList` in protobufs is correct.

Neeharika-Sompalli commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -193,13 +193,14 @@ If the airdrop transaction fees from the list above are covered by a separate pa
+* Verify that an airdrop, which fee expenses are covered by a separate payer account, but has missing payer signature fails
+* Verify that an airdrop with more than 10 entries in the `token_transfers` list fails
+* Verify that an airdrop with a duplicate entry for the same NFT in the `token_transfers` list fails with `PENDING_NFT_AIRDROP_ALREADY_EXISTS`
+* Verify that an airdrop with a duplicate entry for the same fungible token and recipient in the `token_transfers` list works correctly and the fungible token amount is aggregated in the final airdrop to the recipient
```
Is this valid case? In `CryptoTransfer` we reject duplicate entries. We should follow similar pattern imo.

IvanKavaldzhiev replied:
In the case of airdrops, it's allowed for the same recipient to receive a given fungible token multiple times in the same airdrop transaction.

@Nana-EC can you confirm this?

netopyr replied:
I agree with @Neeharika-Sompalli that we should follow the same pattern.

The HIP mentions the case of two separate airdrop transactions for the same token.

IvanKavaldzhiev replied:
Noted

netopyr commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+              - It is an existing one and has `max_auto_associations`, which is a positive number and there are free auto association slots and the recipient is not associated to the token →<br/>
+                Тhe token is auto associated and is directly transferred to the recipient and a `TokenTransferList` is added to the `TransactionRecord`, as well as a new `TokenAssociation` entry in the `automatic_token_associations`
+              - It is an existing one and has `max_auto_associations`, which is a positive number but there are no free auto association slots and the recipient is not associated to the token →<br/>
+                Тhe token transfer is interpreted as a pending and a `TokenPendingAirdrop` entry is added to the `TransactionRecord`. In addition, the airdrop `PendingAirdropId`/`PendingAirdropValue` entry is added to the `PendingAirdrop` VirtualMap in state
```
`VirtualMap` is one possible implementation. In the business logic layer, we should only reference stores.

IvanKavaldzhiev replied:
Fixed

netopyr commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -0,0 +1,206 @@
+
+## Acceptance Tests
+
+* Verify that an airdrop with a token and existing recipient with `maxAutoAssociations=-1` works correctly and the recipient receives the token directly
```
Please add a test that checks the behavior when the feature flag is disabled.

IvanKavaldzhiev replied:
Added

netopyr commented:
The doc does not mention allowances at all. Please explain how allowances may modify the behavior (where applicable).

IvanKavaldzhiev commented:
> The doc does not mention allowances at all. Please explain how allowances may modify the behavior (where applicable).

Added more information about allowances and their use case. Also added test cases related to this scenario.

Neeharika-Sompalli reviewed (approved):
LGTM!

netopyr reviewed (approved):
LGTM - thanks @IvanKavaldzhiev

jsync-swirlds commented on hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md:
```diff
@@ -211,4 +212,5 @@ If the airdrop transaction fees from the list above are covered by a separate pa
 * Verify that an airdrop, which fee expenses are covered by a separate payer account, but has missing payer signature fails
 * Verify that an airdrop with more than 10 entries in the `token_transfers` list fails
 * Verify that an airdrop with a duplicate entry for the same NFT in the `token_transfers` list fails with `PENDING_NFT_AIRDROP_ALREADY_EXISTS`
+* Verify that deletion of an account who has sent an airdrop that is still in pending state fails
```
This probably should not fail (the account owner must be able to delete the account without resorting to a third-party action).  I believe it should delete the pending airdrop instead, so that a pending airdrop cannot "pin" an account in place.

IvanKavaldzhiev replied:
We had some discussions that in order to delete an account that has sent airdrops, you should first delete all related airdrops to this account. So, I believe this should happen with explicit TokenAirdropCancel transactions beforehand.

@Nana-EC, @netopyr  what do you think?

jsync-swirlds replied:
The recipient account cannot cancel the airdrops, so a token holder could, in effect, pin someone else's account by sending airdrops that would have to be claimed (and using a structure that makes claim unpalatable).

jsync-swirlds replied:
I misread the line as a recipient rather than a sender.

After re-reading, I agree with the existing wording.
